pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -971,37 +1028,73 @@ func (p *Pod) FindMatchingWorkloads(ctx context.Context, c client.Client, r reco
 
 	// Cleanup excess pods for each workload pod set (role)
 	activePods := p.runnableOrSucceededPods()
+	inactivePods := p.notRunnableNorSucceededPods()
+
+	var keptPods []corev1.Pod
+	var excessActivePods []corev1.Pod
+	var replacedInactivePods []corev1.Pod
+
 	for _, ps := range workload.Spec.PodSets {
-		// Find all the active pods of the role
-		var roleActivePods []corev1.Pod
-		for _, activePod := range activePods {
-			roleHash, err := getRoleHash(activePod)
+		// Find all the active and inactive pods of the role
+		var roleHashErrors []error
+		hasRoleFunc := func(p *corev1.Pod) bool {
+			hash, err := getRoleHash(*p)
 			if err != nil {
-				return nil, nil, fmt.Errorf("failed to calculate pod role hash: %w", err)
+				roleHashErrors = append(roleHashErrors, err)
```

Comment:
please check the implementation of `errors.Join`

Reply:
In fact, there won't be any additional logic required because we have a minimal number of errors. 

Let's investigate when we encounter `roleHashError`. The `getRoleHash` function only returns an error if `json.Marshal` fails. In our scenario, `json.Marshal` will seldom fail. Therefore, we can use `[]error` or `errors.Join` interchangeably. However, `errors.Join` has advantages - it provides cleaner code and its use will be consistent with the rest of the code.

However, the decision to use `errors.Join` is ultimately up to you.

Reply:
It would be nicer if `errors.Join` appended to the existing `joinError` array instead of creating a new one each time you call it. So for now it seems better to use an array (not just here, but in any case were we need to concatenate errors).